File: fetchData.py
import requests
import os
import csv

api_key_alphavantage = "T9742OP70MN4QXIG"


data_path = "./data/"
symbols_path = "symbols/"
currency_list = "currency_list"
api_key = api_key_alphavantage


def download_currency_list():
    print("Download currency list")
    currency_list_url = "https://www.alphavantage.co/digital_currency_list/"
    r = requests.get(currency_list_url)

    try:
        os.mkdir(data_path)
    except OSError:
        pass
    finally:
        open(data_path + currency_list, "wb").write(r.content)


def loop_currency_list():
    with open(data_path + currency_list) as csv_file:
        csv_reader = csv.reader(csv_file)
        line_count = 0
        for row in csv_reader:
            if line_count > 0:
                download_crypto_datas(row[0])
                print(f'\t{row[0]}')
            line_count += 1

        # print(f'Processed {line_count} lines.')


def download_crypto_datas(currency):
    function = "function=" + "DIGITAL_CURRENCY_MONTHLY"
    apikey = "&apikey=" + api_key
    market = "&market=" + "USD"

    symbols = "&symbol=" + currency

    url = "https://www.alphavantage.co/query?" + function + symbols + market + apikey
    # print(url)
    r = requests.get(url)
    data = r.json()
    try:
        os.mkdir(data_path + symbols_path)
    except OSError:
        pass
    finally:
        open(data_path + symbols_path + currency, "wb").write(r.content)

    print(data)


def fetch_data():
    print("Fetch Data")

    download_currency_list()

    loop_currency_list()

File: IndicielCyrptoMarket/src/main/kotlin/FetchData.kt
import com.google.gson.Gson
import kotlinx.serialization.json.*
import java.io.File
import java.net.URL
import java.nio.file.Files
import java.nio.file.Paths

class FetchData {

    private val dataPath = "./data/"
    private val symbolsPath = "symbols/"
    private val currencyList = "currency_list"
    private val stableCoins =
        arrayListOf("busd", "usdt", "usdc", "dai", "tusd", "usdp", "usdn", "usdd", "fei")

    private val additionalCoins =
        arrayListOf("terra_luna")

    private fun downloadCurrencyList() {
        println("Download currency list")
        val address = "https://api.coingecko.com/api/v3/coins/markets"
        val param1 = "?vs_currency=usd&"
        val param2 = "order=market_cap_desc&"
        val param3 = "include_platform=false"
        val currencyListUrl = address + param1 + param2 + param3

        val r = URL(currencyListUrl).readText()

        Files.createDirectories(Paths.get(dataPath))
        File(dataPath + currencyList).writeText(r)
    }

    private fun loopCurrencyList() {
        val file = File(dataPath + currencyList)
        val json = Json.parseToJsonElement(file.readText()).jsonArray

        for (elem in json) {
            if (stableCoins.contains(elem.jsonObject["symbol"].toString()))
                continue
            downloadCryptoData(elem.jsonObject["id"]!!.jsonPrimitive.content)
        }

        for (elem in additionalCoins)
            downloadCryptoData(elem)
    }

    private fun downloadCryptoData(currency: String) {
        val param1 = "vs_currency=usd&"
        val param2 = "days=max&"
        val param3 = "interval=daily&"

        val url = "https://api.coingecko.com/api/v3/coins/$currency/market_chart/?$param1$param2$param3"
        println(url)

        val r = URL(url).readText()

        Files.createDirectories(Paths.get(dataPath + symbolsPath))

        val purifiedJson = purifyJson(r, currency)
        File(dataPath + symbolsPath + currency).writeText(purifiedJson)
    }

    @Suppress("UNCHECKED_CAST")
    private fun purifyJson(content: String, currency: String): String {
        val j = Gson().fromJson(content, HashMap::class.java)
        j as HashMap<String, Any>





        return ""
    }

    fun fetchData() {
        println("Fetch data")

        File(dataPath).deleteRecursively()

        downloadCurrencyList()

        loopCurrencyList()

    }

}